File: LeetCode/988. Smallest String Starting From Leaf/988.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def smallestFromLeaf(self, root):
        """
        :type root: TreeNode
        :rtype: str
        """
        return self.dfs(root, "")
    
    def dfs(self, node, suffix):
        if not node:
            return None
        
        # Convert the current node's value to the corresponding character
        current_char = chr(node.val + ord('a'))
        # Append to the beginning since we are building from leaf to root
        new_suffix = current_char + suffix
        
        if not node.left and not node.right:  # If it's a leaf node
            return new_suffix
        
        # Recursively find the smallest string from both children
        left_smallest = self.dfs(node.left, new_suffix)
        right_smallest = self.dfs(node.right, new_suffix)
        
        # Determine the lexicographically smallest string
        if left_smallest is None:
            return right_smallest
        if right_smallest is None:
            return left_smallest
        return min(left_smallest, right_smallest)

File: LeetCode/988. Smallest String Starting From Leaf/988.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public String smallestFromLeaf(TreeNode root) {
     return dfs(root, "");
 }

 private String dfs(TreeNode node, String suffix) {
     if (node == null) return null;

     // Convert the current node's value to the corresponding character
     char currentChar = (char) ('a' + node.val);
     // Append to the beginning since we are building from leaf to root
     String newSuffix = currentChar + suffix;

     if (node.left == null && node.right == null) {
         // If it's a leaf, return the current string
         return newSuffix;
     }

     // Recursively find the smallest string from both children
     String leftSmallest = dfs(node.left, newSuffix);
     String rightSmallest = dfs(node.right, newSuffix);

     // Determine the lexicographically smallest string
     if (leftSmallest == null) return rightSmallest;
     if (rightSmallest == null) return leftSmallest;
     return leftSmallest.compareTo(rightSmallest) <= 0 ? leftSmallest : rightSmallest;
 }
}